solr/core/src/java/org/apache/solr/cluster/placement/PlacementPlugin.java

```diff
@@ -0,0 +1,41 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.solr.cluster.placement;
+
+/**
+ * Implemented by external plugins to control replica placement and movement on the search cluster (as well as other things
+ * such as cluster elasticity?) when cluster changes are required (initiated elsewhere, most likely following a Collection
+ * API call).
+ */
+public interface PlacementPlugin {
```

Comment:
I think that we need to add an explicit mechanism for configuration of plugins, otherwise plugin implementors will have to use other Solr facilities anyway.

Maybe add a `configure(Map<String, Object> config)` method?

Reply:
The configuration part I didn't really do yet. Unclear to me how the `configure` method here would be used, since in order to get to it the plugin has to be loaded already...
I was thinking defining the plugin class or classes in some solr configuration file (with the rest of the config). At least a single default plugin implementation that would be used for all placement needs or a default + other ones ones (with names that can then be selected by callers of the Collection API passing a `placement` parameter as suggested in the changes to `CollectionAdminRequest.java`).
Are there similar examples in Solr code (loading plugins possibly on a per collection basis) that I can get inspiration from or reuse?

Reply:
I see a few options for passing config to the plugin:

- Pass the config map in the `computePlacement` call along with the rest of the parameters
- Provide in one of the passed factories the ability of the plugin code to call back into Solr and get the config
- Add a `configure` call on the plugin as you suggest (then the Solr infra decides when to call this method)
- Pass the config when the plugin class is instantiated. This might be equivalent to passing it in the `computePlacement` method if a new plugin class instance is created for each new computation.

So the real choice is do we create a new plugin class instance per placement computation or reuse a given one? Creating a new one for each call is likely a simpler programming model for the plugin developer, it can use class member variables freely and if it wants to keep some state it has to make it static...

With a new instance per computation, there would be no notion of "configuration update". A separate call into the plugin as you suggest or passing the config in `computePlacement` is equivalent (and the latter likely easier to handle in plugin code).

I'm tempted to pass the config with each call to `computePlacement` (assuming the saving of not passing it when the plugin doesn't need it are non measurable). I'm also tempted to make it `Map<String, String>` given it will (most likely?) come from XML and the plugin code would have to deal with what the config means anyway and what types to cast it to...

Reply:
Passing configuration in every request means that the plugin is forced to be stateless. There are of course pros and cons... If we force the stateless model, and plugins has to keep some state, implementors will need to figure out how to do this - static fields are NOT the solution. Maybe pass around some kind of long-lived shared context?

Stateless model makes it easier to reconfigure on the fly for every request ... but how realistic is this scenario? I think reconfiguration is only an occasional event. Furthermore, initialization may be costly so it would make sense to allow plugins to keep some kind of context between invokations.

Reply:
And yeas, in Solr it's usually the responsibility of the parent component to configure its children, at least initially.

Reply:
I believe Solr side would keep the configuration (map) somewhere and just pass a reference to it on every call to the plugin. So cost would be minimal. If the plugin code needs to retrieve a config value, it will basically do a `get` on that map, very low cost here as well.
I don't see how a plugin can keep state between invocations with a semantic that's different from a static field, unless we create a notion of higher level context shared between invocations (and different such higher level contexts will not be shared). I suggest we skip that aspect for now and reconsider later when we implement real plugins. For now I believe the model in which the plugin is instantiated (no arg constructor, or constructor getting the config map or something simple like that) before every call to the computePlacement method would be sufficient.

Reply:
>  get on that map, very low cost here as well
Oh, I wasn't worried about that .. My point was that the plugin may need to initialize costly internal structures based on the config (like fetching some additional params from an external system to eg. verify license constraints or cost schedule for allocating additional resources).

Reply:
I believe we should let the plug-in manage this type of requirements rather than try to control it by the timing of when configs are passed. If there are licences to check the plug-in should cache the result and only confirm each time?

Reply:
Exactly - but where should that data be cached? we don't offer the plugin any way to do it, so implementors will come up with all sorts of nasty ways to circumvent the API sandbox to do it (accessing ZK directly, using external shared FS, or even something worse ;) ).

Reply:
So what do you suggest here and how would the plug-in code deal with it? And how different it would be from the plug-in storing/caching whatever processing it has done in static variables?

I thought about this and we can introduce a concept of "Config session" during which which the config doesn't change.
The plug-in would implement such a newConfigSession method that would accept a config and return an instance of the compute plug-in. That instance would then be called by the framework to do computation. When the framework sees a change of config, it would call again newConfigSession. By using the appropriate classes without storing references in funny places, garbage collection would clean up old configs...

Just having a Solr callback into the existing plug-in passing config is not elegant IMO: forces to synchronize access to config (JMM) and to deal with first compute request coming before first config + race conditions between the two.

Reply:
I'm strongly against using static variables. 

We can either assume that plugins are basically stateless and configured on the fly, or stateful - and in the latter case they could be re-configurable (which has its own problems) or not - that is, you have to create a new instance of the plugin if configuration has changed. In this case there's no problem with synchronization - the old instance of the plugin completes its work using the old config.

Reply:
Perfectly fine with me. The contract will be for the plugin to provide a constructor accepting config to create a new Plugin instance, then that instance will be called for each placement computation (and if the plugin doesn’t care about config, no arg constructor would be fine). Multiple plugin instances might be in use at the same time if so called by Solr (config changes or other reasons).
This means the plugin instance must be reentrant and its member variables can’t directly be used as if specific to a single computation. Not a big deal, plugin implementor can delegate internally to an instance of another class if they want that option.

Reply:
Either constructor convention or - I prefer but not sure how popular it is in the Solr codebase - when configuring a plugin what is added to the Solr config is a plugin factory and that factory is called to get plugin instances. One more interface but cleaner code.